### ISSUE TITLE
Possible Memory Leak

### ISSUE THREAD

**Issue by gopherine** [ripple/rippled#3648]
## Issue Description
Node-size is not definitive, I suspect a possible memory leak, and  I am not particularly sure why, it is it might be because of Sharding. I created a container earlier with 32 GB of ram and 8 cores- setting node size to medium, history shard enabled. This kept on crashing because of over memory usage- from what I understood medium is supposed to consume 16 GB of memory only. Now I increased the ram to 64 GB and the memory usage is roughly around 70% and it keeps on increasing, I would like to know if there is any max cap or should I just decrease the node size. I just feel there is some kind of memory leak somewhere.

## Steps to Reproduce
I used the below container
https://github.com/WietseWind/docker-rippled
```
[server]
port_rpc_admin_local
port_peer

#port_ws_public
# port_ws_admin_local
# ssl_key = /etc/ssl/private/server.key
# ssl_cert = /etc/ssl/certs/server.crt

[port_rpc_admin_local]
port = 5005
ip = 0.0.0.0
admin = 0.0.0.0
protocol = http

#[port_ws_public]
#port = 80
#ip = 0.0.0.0
#protocol = ws

[port_peer]
port = 51235
ip = 0.0.0.0
protocol = peer

# [port_ws_admin_local]
# port = 6006
# ip = 127.0.0.1
# admin = 127.0.0.1
# protocol = ws

[node_size]
medium

# tiny
# small
# medium
# large
# huge

[node_db]
type=NuDB
path=/var/lib/rippled/db/nudb
advisory_delete=0

# How many ledgers do we want to keep (history)?
# Integer value that defines the number of ledgers
# between online deletion events
online_delete=256

# This is the persistent datastore for shards. It is important for the health
# of the ripple network that rippled operators shard as much as practical.
# NuDB requires SSD storage. Helpful information can be found at
# https://xrpl.org/history-sharding.html
[shard_db]
path=/var/lib/rippled/db/shards/nudb
max_size_gb=140

[ledger_history]
# How many ledgers do we want to keep (history)?
# Integer value (ledger count)
# or (if you have lots of TB SSD storage): 'full'
256

[database_path]
/var/lib/rippled/db

[debug_logfile]
/var/log/rippled/debug.log

[sntp_servers]
time.windows.com
time.apple.com
time.nist.gov
pool.ntp.org

[ips]
r.ripple.com 51235

[validators_file]
validators.txt

[rpc_startup]
{ "command": "log_level", "severity": "info" }

# severity (order: lots of information .. only errors)
# debug
# info
# warn
# error
# fatal

[ssl_verify]
1
```

## Expected Result
There should be a cap on memory usage, I believe that should be somewhere around 32 GB

## Actual Result
Memory usage keeps on increasing progressively over the period of time

## Environment
Ubuntu - Running inside docker container

**Comment by mDuo13** [comment]
The `node_size` parameter is not a hard limit on memory usage, which is based on network activity and the overall size of the ledger's state data, among other things. That said, it's not impossible that there is a memory leak here. It's worth looking into, and I think also considering more explicit ways of configuring the server's max memory usage.

**Comment by nbougalis** [comment]
I suspect this is caused by the server downloading and indexing a shard, which is expected. Still, @miguelportilla can you (or someone else) please investigate?

**Comment by gopherine** [comment]
> I suspect this is caused by the server downloading and indexing a shard, which is expected. Still, @miguelportilla can you (or someone else) please investigate?

This is what I suspect too, but still I assume there as to be an upper limit somewhere, Like I am also running a test node, with node size tiny on 32 GB ram and Sharding at 50 GB but there as well the usage has gone over 44% progressively and still increasing. I started this node yesterday. 

For the above today the usage again spiked by 0.6 percent however, as it's increasing the rate of memory spike is getting slow yet still increasing. I will update it again, If at some point its capped. However, I still feel that 16 GB - 48 GB memory usage for a medium size node is a bit too much or either the documentations should be updated respectively. To have a decent idea of range as how much it can grow

**Comment by gopherine** [comment]
@mDuo13 From what I understood, history Sharding enables us, to get data from full history without the need to store full history. However, whenever here I try to get the transaction now from the live server it says transaction not found. And maintains only 256 ledger history as mentioned above

**Comment by mDuo13** [comment]
> update: @mDuo13 From what I understood, history Sharding enables us, to get data from full history without the need to store full history. 

Not quite. Your server only has access to the history in the shards it has downloaded (and its most recent ledger stores). History sharding helps to make overall history available to the network to those who can store it, but there is not currently any mechanism for pulling results "on-demand" from peers in the network.

**Comment by gopherine** [comment]
> > update: @mDuo13 From what I understood, history Sharding enables us, to get data from full history without the need to store full history.
> 
> Not quite. Your server only has access to the history in the shards it has downloaded (and its most recent ledger stores). History sharding helps to make overall history available to the network to those who can store it, but there is not currently any mechanism for pulling results "on-demand" from peers in the network.

Thanks for response, I am not sure but storing full history does not seem to be a cost-efficient solution for ever-growing ledger, And when querying to the public node, It happens quite sometimes, but we get error 'serverTooBusy'. But however I already see a feature request for the same https://github.com/ripple/rippled/issues/2808

--Final Update:: I think the max cap at which this stopped is 72-73% of memory usage however this is still well above 16 GB as mentioned in the doc. And since sharding already stopped , I don't think that this issue is with history-sharding. As mentioned node-size is not hard limit, I understand this, But however there should be a soft cap somewhere near the mentioned node-size this is an issue if a certain ram size is allocated based on it. so if recommended size is 16 GB, the memory usage should not exceed over 32 GB, This helps in having a brief idea on how much resource should be allocated for the same. Especially when deploying on third party containers, This results in unexpected cost and estimations.

**Comment by nbougalis** [comment]
@gopherine: we appreciate the feedback. As you said, having a hard cap is not particularly easy, but it may be reasonable to have the server attempt to monitor its memory usage better.

Considering that this issue was about a memory leak, are you OK if we close this issue since there is no leak? We can open a new issue to improve memory management and resource usage associated with shard processing.

**Comment by leevlad** [comment]
I'm seeing intermittent issues with my `rippled` installation on a 16GB node set at `medium` size. It fails to stay synced up sometimes, which also correlates with severe memory usage increases at the same time. I'm also seeing history sharding in those instances. I should probably mention that I'm running a cluster of these nodes, all of them with exact same specifications and parameters. And I'm seeing some of them stay synced up fine while others are leaking memory struggling to stay synced up.

I'm not sure that a memory leak can be ruled out here. I'd also appreciate any other advice on debugging this.

**Comment by nbougalis** [comment]
Thanks for chiming in @leevlad! Do you have any way of capturing the memory usage of the node over time and the following values from `server_info`: `peer_disconnects`, `peer_disconnects_resources`, `peers` and `server_state`, and putting them on a graph?

**Comment by leevlad** [comment]
Thanks for the suggestion. I noticed that I was actually running on `rippled v1.5`. I've just upgraded my fleet to `rippled v1.6`, and will report back if I see the same behavior over the next few weeks. If I do - I will graph the metrics above and report back.

**Comment by bill-mcgonigle** [comment]
Not sure if this will be helpful ... I landed on this ticket trying to debug what seemed like a memory leak.  Sometime last fall (nightly updates via apt on the official repo) memory usage went a bit nuts, node configured as 'medium' with RocksDB and 'online_delete=2000'.  I had forgotten that I had moved the VM to SSD, so when it recently was spiking to over 11GB and triggering OOM Killer I had systemd limit it to 8/9GB and dug a bit further.  I changed to NuDB, and reading about the memory improvements in 1.7 I changed the node size to 'huge' (1.8.5). It peaked to 3.4GB while rebuilding and is now humming at 2.8GB.  So the main change is offloading RocksDB.  The reason I think there could be a leak is that the system memory charts show a dramatic change after an update last fall and the system didn't change other than nightly package updates.  I don't have a problem currently after switching to NuDB; just wanted to provide the data I have in case it helps others.  You can see the impact of NuDB in the 'weekly' chart below.  The aberrant memory behavior started roughly last September/early October.
![ripple-memory-week](https://user-images.githubusercontent.com/880256/157062434-5e087e1e-83ce-4acf-bafd-bff6de08a36b.png)
![ripple-memory-year](https://user-images.githubusercontent.com/880256/157062464-71b11c49-0bf1-4212-82ca-09abfe0dba2e.png)
